packages/checkbox-group/src/vaadin-checkbox-group.js

```diff
@@ -71,337 +85,272 @@ class CheckboxGroupElement extends ThemableMixin(DirMixin(PolymerElement)) {
-    if (this.value.indexOf(value) === -1) {
-      this.value = this.value.concat(value);
+  __unregisterCheckbox(checkbox) {
+    checkbox.removeEventListener('checked-changed', this.__onCheckboxCheckedChanged);
+
+    if (checkbox.checked) {
+      this.__removeCheckboxFromValue(checkbox.value);
     }
   }
 
   /**
-   * @param {string} value
+   * Override method inherited from `DisabledMixin`
+   * to propagate the `disabled` property to the checkboxes.
+   *
+   * @param {boolean} newValue
+   * @param {boolean} oldValue
+   * @override
    * @protected
    */
-  _removeCheckboxFromValue(value) {
-    this.value = this.value.filter((v) => v !== value);
+  _disabledChanged(newValue, oldValue) {
+    super._disabledChanged(newValue, oldValue);
+
+    // Prevent updating the `disabled` property for the checkboxes at initialization.
+    // Otherwise, the checkboxes may end up enabled regardless the `disabled` attribute
+    // intentionally added by the user on some of them.
+    if (!newValue && oldValue === undefined) {
+      return;
+    }
+
+    if (oldValue !== newValue) {
+      this.__checkboxes.forEach((checkbox) => {
+        checkbox.disabled = newValue;
+      });
+    }
   }
 
   /**
-   * @param {Checkbox} checkbox
-   * @protected
+   * @param {!string} value
+   * @private
    */
-  _changeSelectedCheckbox(checkbox) {
-    if (this._updatingValue) {
+  __addCheckboxToValue(value) {
+    if (this.value.includes(value)) {
       return;
     }
 
+    this.value = [...this.value, value];
+  }
+
+  /**
+   * @param {!string} value
+   * @private
+   */
+  __removeCheckboxFromValue(value) {
+    this.value = this.value.filter((v) => v !== value);
+  }
+
+  /**
+   * @param {!CustomEvent} event
+   * @private
+   */
+  __onCheckboxCheckedChanged(event) {
+    const checkbox = event.target;
+
     if (checkbox.checked) {
-      this._addCheckboxToValue(checkbox.value);
+      this.__addCheckboxToValue(checkbox.value);
     } else {
-      this._removeCheckboxFromValue(checkbox.value);
+      this.__removeCheckboxFromValue(checkbox.value);
     }
   }
 
   /** @private */
-  _updateValue(value) {
+  __valueChanged(value) {
     // setting initial value to empty array, skip validation
     if (value.length === 0 && this._oldValue === undefined) {
       return;
     }
 
-    if (value.length) {
+    if (value.length > 0) {
       this.setAttribute('has-value', '');
     } else {
       this.removeAttribute('has-value');
     }
 
     this._oldValue = value;
-    // set a flag to avoid updating loop
-    this._updatingValue = true;
-    // reflect the value array to checkboxes
-    this._checkboxes.forEach((checkbox) => {
-      checkbox.checked = value.indexOf(checkbox.value) > -1;
+    // // set a flag to avoid updating loop
+    // this._updatingValue = true;
+    // // reflect the value array to checkboxes
+    this.__checkboxes.forEach((checkbox) => {
+      checkbox.checked = value.includes(checkbox.value);
```

Comment:
**TODO:** Make sure that it doesn't cause infinite recursion and, hence, stack overflow.